File: src/macos/LanMessengerTests/H264BitstreamTests.swift
import XCTest
@testable import LanMessenger

// Guards the AVCC <-> Annex-B conversion, which is the biggest cross-platform
// risk in the remote-desktop work: VideoToolbox and Media Foundation package
// H.264 differently, neither decoder accepts the other's packaging, and the
// failure is not a clean error — it is a picture that never appears.
//
// The tests that matter here run against `windows_h264_sample.h264`, a stream
// produced by the real Microsoft H264 Encoder MFT on Dave's Dell
// (spikes/windows-mf-probe, 2026-09-14). Testing a converter against its own
// output proves only that it is self-consistent; testing it against a genuine
// artefact of the other platform's encoder is what actually de-risks this.
final class H264BitstreamTests: XCTestCase {

    // MARK: - Synthetic structure

    private func nal(_ type: UInt8, _ fill: UInt8 = 0xAA, count: Int = 8) -> Data {
        Data([type] + [UInt8](repeating: fill, count: count))
    }

    func testScannerConsumesFourByteStartCodesWhole() {
        // The bug this pins: a 4-byte start code contains a 3-byte one at
        // offset+1, so a scanner that advances a single byte finds a phantom NAL
        // inside every real one and reports exactly double. That is not
        // hypothetical — the first analysis of the Windows fixture reported
        // "126 four-byte and 126 three-byte codes", and the suspiciously equal
        // counts were the only clue.
        var stream = Data()
        stream.append(contentsOf: [0, 0, 0, 1]); stream.append(nal(H264NALType.sps))
        stream.append(contentsOf: [0, 0, 0, 1]); stream.append(nal(H264NALType.pps))
        stream.append(contentsOf: [0, 0, 0, 1]); stream.append(nal(H264NALType.idrSlice))

        let units = H264Bitstream.scanAnnexB(stream)
        XCTAssertEqual(units.count, 3, "a 4-byte start code must be consumed whole")
        XCTAssertEqual(units.map { $0.type },
                       [H264NALType.sps, H264NALType.pps, H264NALType.idrSlice])
    }

    func testScannerHandlesThreeByteStartCodes() {
        var stream = Data()
        stream.append(contentsOf: [0, 0, 1]); stream.append(nal(H264NALType.sps))
        stream.append(contentsOf: [0, 0, 1]); stream.append(nal(H264NALType.idrSlice))
        let units = H264Bitstream.scanAnnexB(stream)
        XCTAssertEqual(units.count, 2)
        XCTAssertEqual(units.map { $0.type }, [H264NALType.sps, H264NALType.idrSlice])
    }

    func testScannerHandlesMixedStartCodeSizes() {
        // Legal, and some encoders do it. Our own output is uniformly 4-byte,
        // but the reader must not assume the peer's is.
        var stream = Data()
        stream.append(contentsOf: [0, 0, 0, 1]); stream.append(nal(H264NALType.sps))
        stream.append(contentsOf: [0, 0, 1]);    stream.append(nal(H264NALType.pps))
        stream.append(contentsOf: [0, 0, 0, 1]); stream.append(nal(H264NALType.idrSlice))
        XCTAssertEqual(H264Bitstream.scanAnnexB(stream).map { $0.type },
                       [H264NALType.sps, H264NALType.pps, H264NALType.idrSlice])
    }

    func testEmptyAndStartCodeOnlyBuffers() {
        XCTAssertTrue(H264Bitstream.scanAnnexB(Data()).isEmpty)
        XCTAssertTrue(H264Bitstream.scanAnnexB(Data([0, 0, 0, 1])).isEmpty,
                      "a start code with no payload is not a NAL unit")
        XCTAssertThrowsError(try H264Bitstream.annexBToAVCC(Data([1, 2, 3])))
    }

    // MARK: - Round trips

    func testAnnexBToAVCCAndBack() throws {
        let slice = nal(H264NALType.nonIDRSlice, 0x11, count: 40)
        let stream = H264Bitstream.annexB(from: [slice])

        let avcc = try H264Bitstream.annexBToAVCC(stream)
        XCTAssertEqual(avcc.count, 4 + slice.count, "4-byte length prefix plus payload")
        XCTAssertEqual(Array(avcc.prefix(4)), [0, 0, 0, UInt8(slice.count)])

        let back = try H264Bitstream.avccToAnnexB(avcc, nalLengthSize: 4)
        XCTAssertEqual(back, stream)
    }

    func testAVCCLengthSizesOtherThanFour() throws {
        let slice = nal(H264NALType.nonIDRSlice, 0x22, count: 20)
        let stream = H264Bitstream.annexB(from: [slice])
        for size in [1, 2, 4] {
            let avcc = try H264Bitstream.annexBToAVCC(stream, nalLengthSize: size)
            XCTAssertEqual(avcc.count, size + slice.count)
            let units = try H264Bitstream.scanAVCC(avcc, nalLengthSize: size)
            XCTAssertEqual(units.count, 1)
            XCTAssertEqual(units[0].payload, slice)
        }
        XCTAssertThrowsError(try H264Bitstream.annexBToAVCC(stream, nalLengthSize: 3),
                             "3 is not a legal NAL length size")
    }

    func testParameterSetsAreDroppedFromSampleData() throws {
        // SPS/PPS belong in the CMVideoFormatDescription. Leaving them in the
        // sample data is a decode error on VideoToolbox, not a harmless
        // duplicate.
        let stream = H264Bitstream.annexB(from: [
            nal(H264NALType.accessUnitDelimiter, 0x01, count: 1),
            nal(H264NALType.sps, 0x67, count: 12),
            nal(H264NALType.pps, 0x68, count: 3),
            nal(H264NALType.idrSlice, 0x65, count: 60),
        ])
        let avcc = try H264Bitstream.annexBToAVCC(stream)
        let units = try H264Bitstream.scanAVCC(avcc, nalLengthSize: 4)
        XCTAssertEqual(units.map { $0.type }, [H264NALType.idrSlice],
                       "only picture data belongs in the sample")
    }

    func testKeyframeGetsParameterSetsPrependedOnTheWayOut() throws {
        // Media Foundation has no out-of-band channel for parameter sets, so a
        // stream whose SPS/PPS live only in our format description decodes to
        // nothing on Windows.
        let sps = nal(H264NALType.sps, 0x67, count: 12)
        let pps = nal(H264NALType.pps, 0x68, count: 3)
        let idr = nal(H264NALType.idrSlice, 0x65, count: 50)
        let avcc = try H264Bitstream.annexBToAVCC(H264Bitstream.annexB(from: [idr]))

        let out = try H264Bitstream.avccToAnnexB(avcc, nalLengthSize: 4,
                                                 parameterSets: (sps: [sps], pps: [pps]))
        XCTAssertEqual(H264Bitstream.scanAnnexB(out).map { $0.type },
                       [H264NALType.sps, H264NALType.pps, H264NALType.idrSlice])
    }

    func testNonKeyframeDoesNotGetParameterSets() throws {
        let sps = nal(H264NALType.sps, 0x67, count: 12)
        let pps = nal(H264NALType.pps, 0x68, count: 3)
        let slice = nal(H264NALType.nonIDRSlice, 0x41, count: 50)
        let avcc = try H264Bitstream.annexBToAVCC(H264Bitstream.annexB(from: [slice]))

        let out = try H264Bitstream.avccToAnnexB(avcc, nalLengthSize: 4,
                                                 parameterSets: (sps: [sps], pps: [pps]))
        XCTAssertEqual(H264Bitstream.scanAnnexB(out).map { $0.type }, [H264NALType.nonIDRSlice],
                       "repeating parameter sets on every frame is pure bitrate waste")
    }

    // MARK: - Malformed AVCC

    func testMalformedAVCCIsRejectedNotMisparsed() {
        // A hostile or corrupt length prefix must fault rather than walk off the
        // end of the buffer.
        var claimsTooMuch = Data([0x00, 0x00, 0xFF, 0xFF])
        claimsTooMuch.append(Data(repeating: 0x41, count: 10))
        XCTAssertThrowsError(try H264Bitstream.scanAVCC(claimsTooMuch, nalLengthSize: 4))

        XCTAssertThrowsError(try H264Bitstream.scanAVCC(Data([0x00, 0x00]), nalLengthSize: 4),
                             "a truncated length prefix must fault")
        XCTAssertThrowsError(try H264Bitstream.scanAVCC(Data([0, 0, 0, 0]), nalLengthSize: 4),
                             "a zero-length NAL would loop forever")
    }

    // MARK: - The real Windows encoder output

    private func windowsSample() throws -> Data {
        let url: URL
        if let bundled = Bundle(for: H264BitstreamTests.self)
            .url(forResource: "windows_h264_sample", withExtension: "h264") {
            url = bundled
        } else {
            url = URL(fileURLWithPath: #file).deletingLastPathComponent()
                .appendingPathComponent("windows_h264_sample.h264")
        }
        return try Data(contentsOf: url)
    }

    func testWindowsSampleStructureMatchesWhatTheEncoderActuallyProduced() throws {
        let data = try windowsSample()
        let units = H264Bitstream.scanAnnexB(data)

        // Pinned against the recorded run: 60 NV12 frames in, 126 NAL units out.
        XCTAssertEqual(units.count, 126, "scanner disagrees with the recorded structure")

        var counts: [UInt8: Int] = [:]
        for unit in units { counts[unit.type, default: 0] += 1 }
        XCTAssertEqual(counts[H264NALType.accessUnitDelimiter], 60, "one AUD per frame")
        XCTAssertEqual(counts[H264NALType.nonIDRSlice], 58)
        XCTAssertEqual(counts[H264NALType.idrSlice], 2)
        XCTAssertEqual(counts[H264NALType.sps], 2)
        XCTAssertEqual(counts[H264NALType.pps], 2)
        XCTAssertEqual((counts[H264NALType.nonIDRSlice] ?? 0) + (counts[H264NALType.idrSlice] ?? 0), 60,
                       "one slice per frame")
    }

    func testWindowsSampleCarriesParameterSetsInBandBeforeEveryIDR() throws {
        // This is the property the macOS receive path depends on. If Media
        // Foundation ever stopped doing it we would have to negotiate parameter
        // sets out of band, so it is worth a test rather than an assumption.
        let units = H264Bitstream.scanAnnexB(try windowsSample())
        let idrIndices = units.indices.filter { units[$0].isKeyframe }
        XCTAssertEqual(idrIndices.count, 2)

        for idr in idrIndices {
            // Look back to the start of the access unit, not a fixed number of
            // NAL units: the encoder puts two SEI messages between the parameter
            // sets and the first IDR, so a narrow window misses them. The access
            // unit delimiter is the real boundary.
            var start = idr
            while start > 0 && units[start - 1].type != H264NALType.accessUnitDelimiter {
                start -= 1
            }
            let accessUnit = units[start..<idr].map { $0.type }
            XCTAssertTrue(accessUnit.contains(H264NALType.sps),
                          "no SPS in the access unit containing the IDR at \(idr), saw \(accessUnit)")
            XCTAssertTrue(accessUnit.contains(H264NALType.pps),
                          "no PPS in the access unit containing the IDR at \(idr), saw \(accessUnit)")
        }
    }

    func testWindowsSampleParameterSetsAreExtractable() throws {
        let (sps, pps) = H264Bitstream.parameterSets(inAnnexB: try windowsSample())
        XCTAssertEqual(sps.count, 2)
        XCTAssertEqual(pps.count, 2)
        // Recorded sizes from the real stream; a converter that silently
        // truncated a parameter set would still "work" without this.
        XCTAssertEqual(sps[0].count, 25)
        XCTAssertEqual(pps[0].count, 4)
        XCTAssertEqual(sps[0][sps[0].startIndex] & 0x1F, H264NALType.sps)
        XCTAssertEqual(pps[0][pps[0].startIndex] & 0x1F, H264NALType.pps)
        XCTAssertEqual(sps[0], sps[1], "both IDRs should carry identical parameter sets")
    }

    func testWindowsSampleConvertsToAVCCAndBackWithoutLoss() throws {
        // The end-to-end property: Windows Annex-B in, AVCC for VideoToolbox,
        // Annex-B back out for the return path, with every picture NAL intact.
        let original = try windowsSample()
        let pictureUnits = H264Bitstream.scanAnnexB(original)
            .filter { !H264Bitstream.nonPictureTypes.contains($0.type) }

        let avcc = try H264Bitstream.annexBToAVCC(original)
        let roundTripped = try H264Bitstream.scanAVCC(avcc, nalLengthSize: 4)

        XCTAssertEqual(roundTripped.count, pictureUnits.count)
        XCTAssertEqual(roundTripped.map { $0.payload }, pictureUnits.map { $0.payload },
                       "payload bytes must survive the conversion untouched")

        let back = try H264Bitstream.avccToAnnexB(avcc, nalLengthSize: 4)
        XCTAssertEqual(H264Bitstream.scanAnnexB(back).map { $0.payload },
                       pictureUnits.map { $0.payload })
    }

    func testWindowsSampleKeyframeDetection() throws {
        let data = try windowsSample()
        XCTAssertTrue(H264Bitstream.annexBContainsKeyframe(data))

        // And a slice-only excerpt must not be mistaken for one.
        let nonIDR = H264Bitstream.scanAnnexB(data).first { $0.type == H264NALType.nonIDRSlice }
        let excerpt = H264Bitstream.annexB(from: [try XCTUnwrap(nonIDR).payload])
        XCTAssertFalse(H264Bitstream.annexBContainsKeyframe(excerpt))
    }
}

File: src/windows-native/LanMessenger.Tests/H264BitstreamTests.cs
using LanMessenger.Core.Networking.Media;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace LanMessenger.Tests;

// Guards the AVCC <-> Annex-B conversion. Mirror of the macOS
// H264BitstreamTests.swift, with the same test names so a failure here names a
// macOS test that already passes.
//
// The tests that matter run against windows_h264_sample.h264, produced by the
// real Microsoft H264 Encoder MFT on this very machine
// (spikes/windows-mf-probe, 2026-09-14). Testing a converter against its own
// output proves only self-consistency; testing it against a genuine encoder
// artefact is what de-risks the interop.
[TestClass]
public class H264BitstreamTests
{
    private static byte[] Nal(byte type, byte fill = 0xAA, int count = 8)
    {
        var d = new byte[count + 1];
        d[0] = type;
        for (int i = 1; i < d.Length; i++) d[i] = fill;
        return d;
    }

    // ---- Synthetic structure ----------------------------------------------

    [TestMethod]
    public void ScannerConsumesFourByteStartCodesWhole()
    {
        // The bug this pins: a 4-byte start code contains a 3-byte one at
        // offset+1, so a scanner that advances a single byte finds a phantom NAL
        // inside every real one and reports exactly double. Not hypothetical —
        // the first analysis of this very fixture reported "126 four-byte and
        // 126 three-byte codes", and the equal counts were the only clue.
        byte[] stream = H264Bitstream.AnnexB(
            Nal(H264NALType.SPS), Nal(H264NALType.PPS), Nal(H264NALType.IDRSlice));

        var units = H264Bitstream.ScanAnnexB(stream);
        Assert.AreEqual(3, units.Count, "a 4-byte start code must be consumed whole");
        CollectionAssert.AreEqual(
            new byte[] { H264NALType.SPS, H264NALType.PPS, H264NALType.IDRSlice },
            units.Select(u => u.Type).ToArray());
    }

    [TestMethod]
    public void ScannerHandlesThreeByteStartCodes()
    {
        var stream = new List<byte>();
        stream.AddRange([0, 0, 1]); stream.AddRange(Nal(H264NALType.SPS));
        stream.AddRange([0, 0, 1]); stream.AddRange(Nal(H264NALType.IDRSlice));
        var units = H264Bitstream.ScanAnnexB(stream.ToArray());
        Assert.AreEqual(2, units.Count);
        CollectionAssert.AreEqual(new byte[] { H264NALType.SPS, H264NALType.IDRSlice },
                                  units.Select(u => u.Type).ToArray());
    }

    [TestMethod]
    public void ScannerHandlesMixedStartCodeSizes()
    {
        // Legal, and some encoders do it. Our own output is uniformly 4-byte,
        // but the reader must not assume the peer's is.
        var stream = new List<byte>();
        stream.AddRange([0, 0, 0, 1]); stream.AddRange(Nal(H264NALType.SPS));
        stream.AddRange([0, 0, 1]);    stream.AddRange(Nal(H264NALType.PPS));
        stream.AddRange([0, 0, 0, 1]); stream.AddRange(Nal(H264NALType.IDRSlice));
        CollectionAssert.AreEqual(
            new byte[] { H264NALType.SPS, H264NALType.PPS, H264NALType.IDRSlice },
            H264Bitstream.ScanAnnexB(stream.ToArray()).Select(u => u.Type).ToArray());
    }

    [TestMethod]
    public void EmptyAndStartCodeOnlyBuffers()
    {
        Assert.AreEqual(0, H264Bitstream.ScanAnnexB([]).Count);
        Assert.AreEqual(0, H264Bitstream.ScanAnnexB([0, 0, 0, 1]).Count,
            "a start code with no payload is not a NAL unit");
        Assert.ThrowsException<H264BitstreamException>(() => H264Bitstream.AnnexBToAVCC([1, 2, 3]));
    }

    // ---- Round trips -------------------------------------------------------

    [TestMethod]
    public void AnnexBToAVCCAndBack()
    {
        byte[] slice = Nal(H264NALType.NonIDRSlice, 0x11, 40);
        byte[] stream = H264Bitstream.AnnexB(slice);

        byte[] avcc = H264Bitstream.AnnexBToAVCC(stream);
        Assert.AreEqual(4 + slice.Length, avcc.Length, "4-byte length prefix plus payload");
        CollectionAssert.AreEqual(new byte[] { 0, 0, 0, (byte)slice.Length }, avcc[..4]);

        CollectionAssert.AreEqual(stream, H264Bitstream.AVCCToAnnexB(avcc, 4));
    }

    [TestMethod]
    public void AVCCLengthSizesOtherThanFour()
    {
        byte[] slice = Nal(H264NALType.NonIDRSlice, 0x22, 20);
        byte[] stream = H264Bitstream.AnnexB(slice);
        foreach (int size in new[] { 1, 2, 4 })
        {
            byte[] avcc = H264Bitstream.AnnexBToAVCC(stream, size);
            Assert.AreEqual(size + slice.Length, avcc.Length);
            var units = H264Bitstream.ScanAVCC(avcc, size);
            Assert.AreEqual(1, units.Count);
            CollectionAssert.AreEqual(slice, units[0].Payload);
        }
        Assert.ThrowsException<H264BitstreamException>(() => H264Bitstream.AnnexBToAVCC(stream, 3),
            "3 is not a legal NAL length size");
    }

    [TestMethod]
    public void ParameterSetsAreDroppedFromSampleData()
    {
        byte[] stream = H264Bitstream.AnnexB(
            Nal(H264NALType.AccessUnitDelimiter, 0x01, 1),
            Nal(H264NALType.SPS, 0x67, 12),
            Nal(H264NALType.PPS, 0x68, 3),
            Nal(H264NALType.IDRSlice, 0x65, 60));
        var units = H264Bitstream.ScanAVCC(H264Bitstream.AnnexBToAVCC(stream), 4);
        CollectionAssert.AreEqual(new byte[] { H264NALType.IDRSlice },
            units.Select(u => u.Type).ToArray(), "only picture data belongs in the sample");
    }

    [TestMethod]
    public void KeyframeGetsParameterSetsPrependedOnTheWayOut()
    {
        byte[] sps = Nal(H264NALType.SPS, 0x67, 12);
        byte[] pps = Nal(H264NALType.PPS, 0x68, 3);
        byte[] avcc = H264Bitstream.AnnexBToAVCC(H264Bitstream.AnnexB(Nal(H264NALType.IDRSlice, 0x65, 50)));

        byte[] outBytes = H264Bitstream.AVCCToAnnexB(avcc, 4, [sps], [pps]);
        CollectionAssert.AreEqual(
            new byte[] { H264NALType.SPS, H264NALType.PPS, H264NALType.IDRSlice },
            H264Bitstream.ScanAnnexB(outBytes).Select(u => u.Type).ToArray());
    }

    [TestMethod]
    public void NonKeyframeDoesNotGetParameterSets()
    {
        byte[] sps = Nal(H264NALType.SPS, 0x67, 12);
        byte[] pps = Nal(H264NALType.PPS, 0x68, 3);
        byte[] avcc = H264Bitstream.AnnexBToAVCC(H264Bitstream.AnnexB(Nal(H264NALType.NonIDRSlice, 0x41, 50)));

        byte[] outBytes = H264Bitstream.AVCCToAnnexB(avcc, 4, [sps], [pps]);
        CollectionAssert.AreEqual(new byte[] { H264NALType.NonIDRSlice },
            H264Bitstream.ScanAnnexB(outBytes).Select(u => u.Type).ToArray(),
            "repeating parameter sets on every frame is pure bitrate waste");
    }

    // ---- Malformed AVCC ----------------------------------------------------

    [TestMethod]
    public void MalformedAVCCIsRejectedNotMisparsed()
    {
        var claimsTooMuch = new List<byte> { 0x00, 0x00, 0xFF, 0xFF };
        claimsTooMuch.AddRange(Enumerable.Repeat((byte)0x41, 10));
        Assert.ThrowsException<H264BitstreamException>(
            () => H264Bitstream.ScanAVCC(claimsTooMuch.ToArray(), 4));

        Assert.ThrowsException<H264BitstreamException>(
            () => H264Bitstream.ScanAVCC([0x00, 0x00], 4), "a truncated length prefix must fault");
        Assert.ThrowsException<H264BitstreamException>(
            () => H264Bitstream.ScanAVCC([0, 0, 0, 0], 4), "a zero-length NAL would loop forever");
    }

    // ---- The real Windows encoder output -----------------------------------

    private static byte[] WindowsSample()
    {
        const string name = "windows_h264_sample.h264";
        if (File.Exists(name)) return File.ReadAllBytes(name);
        string beside = Path.Combine(AppContext.BaseDirectory, name);
        if (File.Exists(beside)) return File.ReadAllBytes(beside);
        return File.ReadAllBytes(Path.Combine(
            Path.GetDirectoryName(typeof(H264BitstreamTests).Assembly.Location) ?? ".", name));
    }

    [TestMethod]
    public void WindowsSampleStructureMatchesWhatTheEncoderActuallyProduced()
    {
        var units = H264Bitstream.ScanAnnexB(WindowsSample());

        // Pinned against the recorded run: 60 NV12 frames in, 126 NAL units out.
        Assert.AreEqual(126, units.Count, "scanner disagrees with the recorded structure");

        var counts = units.GroupBy(u => u.Type).ToDictionary(g => g.Key, g => g.Count());
        Assert.AreEqual(60, counts[H264NALType.AccessUnitDelimiter], "one AUD per frame");
        Assert.AreEqual(58, counts[H264NALType.NonIDRSlice]);
        Assert.AreEqual(2,  counts[H264NALType.IDRSlice]);
        Assert.AreEqual(2,  counts[H264NALType.SPS]);
        Assert.AreEqual(2,  counts[H264NALType.PPS]);
        Assert.AreEqual(60, counts[H264NALType.NonIDRSlice] + counts[H264NALType.IDRSlice],
            "one slice per frame");
    }

    [TestMethod]
    public void WindowsSampleCarriesParameterSetsInBandBeforeEveryIDR()
    {
        // The property the macOS receive path depends on. If Media Foundation
        // ever stopped doing it we would have to negotiate parameter sets out of
        // band, so it is worth a test rather than an assumption.
        var units = H264Bitstream.ScanAnnexB(WindowsSample());
        var idrIndices = Enumerable.Range(0, units.Count).Where(i => units[i].IsKeyframe).ToList();
        Assert.AreEqual(2, idrIndices.Count);

        foreach (int idr in idrIndices)
        {
            // Look back to the start of the access unit, not a fixed number of
            // NAL units: the encoder puts two SEI messages between the parameter
            // sets and the first IDR, so a narrow window misses them.
            int start = idr;
            while (start > 0 && units[start - 1].Type != H264NALType.AccessUnitDelimiter) start--;
            var accessUnit = units.GetRange(start, idr - start).Select(u => u.Type).ToList();
            Assert.IsTrue(accessUnit.Contains(H264NALType.SPS),
                $"no SPS in the access unit containing the IDR at {idr}");
            Assert.IsTrue(accessUnit.Contains(H264NALType.PPS),
                $"no PPS in the access unit containing the IDR at {idr}");
        }
    }

    [TestMethod]
    public void WindowsSampleParameterSetsAreExtractable()
    {
        var (sps, pps) = H264Bitstream.ParameterSets(WindowsSample());
        Assert.AreEqual(2, sps.Count);
        Assert.AreEqual(2, pps.Count);
        // Recorded sizes from the real stream; a converter that silently
        // truncated a parameter set would still "work" without this.
        Assert.AreEqual(25, sps[0].Length);
        Assert.AreEqual(4,  pps[0].Length);
        Assert.AreEqual(H264NALType.SPS, (byte)(sps[0][0] & 0x1F));
        Assert.AreEqual(H264NALType.PPS, (byte)(pps[0][0] & 0x1F));
        CollectionAssert.AreEqual(sps[0], sps[1], "both IDRs should carry identical parameter sets");
    }

    [TestMethod]
    public void WindowsSampleConvertsToAVCCAndBackWithoutLoss()
    {
        // The end-to-end property: Windows Annex-B in, AVCC for VideoToolbox,
        // Annex-B back out for the return path, with every picture NAL intact.
        byte[] original = WindowsSample();
        var pictureUnits = H264Bitstream.ScanAnnexB(original)
            .Where(u => !H264Bitstream.NonPictureTypes.Contains(u.Type)).ToList();

        byte[] avcc = H264Bitstream.AnnexBToAVCC(original);
        var roundTripped = H264Bitstream.ScanAVCC(avcc, 4);

        Assert.AreEqual(pictureUnits.Count, roundTripped.Count);
        for (int i = 0; i < pictureUnits.Count; i++)
            CollectionAssert.AreEqual(pictureUnits[i].Payload, roundTripped[i].Payload,
                $"payload bytes of NAL {i} must survive the conversion untouched");

        var back = H264Bitstream.ScanAnnexB(H264Bitstream.AVCCToAnnexB(avcc, 4));
        Assert.AreEqual(pictureUnits.Count, back.Count);
    }

    [TestMethod]
    public void WindowsSampleKeyframeDetection()
    {
        byte[] data = WindowsSample();
        Assert.IsTrue(H264Bitstream.AnnexBContainsKeyframe(data));

        var nonIDR = H264Bitstream.ScanAnnexB(data).First(u => u.Type == H264NALType.NonIDRSlice);
        Assert.IsFalse(H264Bitstream.AnnexBContainsKeyframe(H264Bitstream.AnnexB(nonIDR.Payload)));
    }

    // ---- The macOS encoder output, for the other direction ------------------

    private static byte[] MacOSSample()
    {
        var path = Path.Combine(AppContext.BaseDirectory, "macos_h264_sample.h264");
        if (!File.Exists(path)) path = "macos_h264_sample.h264";
        return File.ReadAllBytes(path);
    }

    [TestMethod]
    public void MacOSSampleIsTheExactShapeTheWindowsOneIsNot()
    {
        // Real VideoToolbox output, committed so this suite can assert the
        // macOS -> Windows direction without a Mac, exactly as the Swift suite
        // asserts the reverse without a PC.
        //
        // Its value is in how unlike the Windows sample it is: VideoToolbox
        // emits no access unit delimiters at all and parameter sets only at
        // IDRs. It is direct evidence for the rule that an AUD-based split
        // collapses a stream — there are no AUDs here to split on.
        var units = H264Bitstream.ScanAnnexB(MacOSSample());
        Assert.AreEqual(64, units.Count, "pinned against the recorded run");

        var counts = new Dictionary<byte, int>();
        foreach (var unit in units)
        {
            counts.TryGetValue(unit.Type, out var n);
            counts[unit.Type] = n + 1;
        }

        Assert.IsFalse(counts.ContainsKey(H264NALType.AccessUnitDelimiter),
                       "VideoToolbox emits no AUDs — that is the point of this fixture");
        Assert.AreEqual(2, counts[H264NALType.IDRSlice]);
        Assert.AreEqual(58, counts[H264NALType.NonIDRSlice]);
        Assert.AreEqual(2, counts[H264NALType.SPS], "parameter sets ride the IDRs");
        Assert.AreEqual(2, counts[H264NALType.PPS]);
    }

    [TestMethod]
    public void MacOSSampleCarriesParameterSetsInBandBeforeEveryIDR()
    {
        // What Media Foundation depends on: it has no out-of-band channel for
        // parameter sets, so a stream whose SPS/PPS live only in a
        // CMVideoFormatDescription decodes to nothing here.
        var units = H264Bitstream.ScanAnnexB(MacOSSample());
        for (int i = 0; i < units.Count; i++)
        {
            if (units[i].Type != H264NALType.IDRSlice) continue;
            bool sps = false, pps = false;
            for (int k = Math.Max(0, i - 4); k < i; k++)
            {
                if (units[k].Type == H264NALType.SPS) sps = true;
                if (units[k].Type == H264NALType.PPS) pps = true;
            }
            Assert.IsTrue(sps, $"IDR at {i} has no SPS before it");
            Assert.IsTrue(pps, $"IDR at {i} has no PPS before it");
        }
    }
}
